include/UniqueTable.hpp

```diff
@@ -0,0 +1,353 @@
+                                it = ++lastit;
+                            }
+                            collected++;
+                        } else {
+                            ++it;
+                            ++lastit;
+                            remaining++;
+                        }
+                    }
+                }
+            }
+            gcLimit += gcIncrement;
+            nodeCount = remaining;
+            return collected;
+        }
+
+        void clear() {
+            // clear unique table buckets
+            for (auto& table: tables) {
+                for (auto& bucket: table) {
+                    bucket.clear();
+                }
+            }
+            // clear available stack
+            while (!available.empty())
+                available.pop();
+
+            // release memory of all but the first chunk TODO: it could be desirable to keep the memory
+            while (chunkID > 0) {
+                chunks.pop_back();
+                chunkID--;
+            }
+            // restore initial chunk setting
+            chunkIt        = chunks[0].begin();
+            chunkEndIt     = chunks[0].end();
+            allocationSize = INITIAL_ALLOCATION_SIZE * growthFactor;
+            allocations    = INITIAL_ALLOCATION_SIZE;
+
+            nodeCount     = 0;
+            peakNodeCount = 0;
+
+            collisions = 0;
+            hits       = 0;
+            lookups    = 0;
+
+            std::fill(active.begin(), active.end(), 0);
+            activeNodeCount = 0;
+            maxActive       = 0;
+
+            gcCalls = 0;
+            gcRuns  = 0;
+            gcLimit = gcInitialLimit;
+        };
+
+        void print() {
+            Qubit q = nvars - 1;
+            for (auto it = tables.rbegin(); it != tables.rend(); ++it) {
+                auto& table = *it;
+                std::cout << "\t" << static_cast<std::size_t>(q) << ":" << std::endl;
+                for (size_t key = 0; key < table.size(); ++key) {
+                    auto& bucket = table[key];
+                    if (!bucket.empty())
+                        std::cout << key << ": ";
+
+                    for (const auto& node: bucket)
+                        std::cout << "\t\t" << reinterpret_cast<uintptr_t>(node) << " " << node->ref << "\t";
+
+                    if (!bucket.empty())
+                        std::cout << std::endl;
+                }
+                --q;
+            }
+        }
+
+        void printActive() {
+            std::cout << "#printActive: " << activeNodeCount << ", ";
+            for (const auto& a: active)
+                std::cout << a << " ";
+            std::cout << std::endl;
+        }
+
+        [[nodiscard]] fp hitRatio() const { return static_cast<fp>(hits) / lookups; }
+        [[nodiscard]] fp colRatio() const { return static_cast<fp>(collisions) / lookups; }
+
+        [[nodiscard]] auto getActiveNodeCount() const {
```

Comment:
I'd prefer to write `std::size_t` instead of `auto` for clarity. Same for `getActiveNodeCount(Qubit)`.

Reply:
Agreed 👍🏼